### PR TITLE
Unify patching / tiling 

### PR DESCRIPTION
**Pull request by mh-nguyen712**
Toward closing #1039

In this PR:
- refactor the TiledMixin2d to have functional API, which can be used to define `deepinv.utils.image_to_patches` and `deepinv.utils.patches_to_image`, the function `deepinv.utils.patchify` now becomes an alias of `deepinv.utils.image_to_patches`. Both the mixin and the functional API use the same implementation under the hood.
- This new implementation is then used in:
    + `PatchDataset` with `TiledMixin2d` 
    + `PatchCovarianceNoiseEstimator` with  `deepinv.utils.image_to_patches`

Remaining issues in #1039: 
- PatchSampler: the current TiledMixin2d and functions are not optimal for this, since we need random patches (different than fixed stride patches)
- DistributedSampler: this would require more works, since this module is still in active dev, eg #1088 

 

Thank you for contributing to DeepInverse!

Please refer to our [contributing guidelines](https://deepinv.github.io/deepinv/contributing.html) for full instructions on how to contribute, including writing tests, documentation and code style.

Once the GitHub tests have been approved by a maintainer (only required for first-time contributors), and the `Build Docs` GitHub action has run successfully, you can download the documentation as a zip file from the [Actions page](https://github.com/deepinv/deepinv/actions/workflows/documentation.yml). Look for the workflow run corresponding to your pull request.

### Checks to be done before submitting your PR

- [ ] `python3 -m pytest deepinv/tests` runs successfully.
- [ ] `black .` and `ruff check .` run successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### CODE REVIEW

**Comment by Andrewwango** [comment]
Really nice contribution @mh-nguyen712  thank you! Please can you add a PR description with what you did?

**Comment by mh-nguyen712** [comment]
Thanks @Andrewwango, added!

File: deepinv/utils/patch_extractor.py [hunk 1]
```diff
@@ -56,3 +74,100 @@ def patch_extractor(
...
+        stride=stride_2d,
+        pad_if_needed=pad_if_needed,
+    )
+
+
+def patches_to_image(
+    patches: Tensor,
+    stride: int | tuple[int, int],
+    img_size: tuple[int, int] | None = None,
+    reduce_overlap: str = "sum",
```
Comment (romainvo) [inline deepinv/utils/patch_extractor.py]: I would set the default value to "mean" otherwise it changes the pixel range of the reconstructed image

File: deepinv/utils/_tiling.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        3, patch_size[1], stride[1]
+    )
+    return patches.contiguous()
+
+
+def _patches_to_image_impl(
+    patches: Tensor,
+    stride: tuple[int, int],
+    img_size: tuple[int, int] | None = None,
+    reduce_overlap: str = "sum",
```
Comment (romainvo) [inline deepinv/utils/_tiling.py]: I would set the default value to "mean" otherwise it changes the pixel range of the reconstructed image. (and set reduce_overlap="mean" across all the methods calling it)

File: deepinv/utils/patch_extractor.py [hunk 3]
```diff
@@ -56,3 +74,100 @@ def patch_extractor(
...
+    The behavior mirrors :class:`deepinv.utils.TiledMixin2d` while exposing a
+    functional API.
+
+    :param torch.Tensor image: Input image tensor of shape ``(B, C, H, W)``.
+    :param int | tuple[int, int] patch_size: Patch size ``(patch_h, patch_w)``.
+        If an ``int`` is provided, the same value is used for both dimensions.
+    :param int | tuple[int, int] stride: Stride between adjacent patches as
+        ``(stride_h, stride_w)``. If ``None``, defaults to half the patch size.
+    :param bool pad_if_needed: If ``True``, adds extra right/bottom padding so
+        that all extracted patches have the requested ``patch_size``.
```
Comment (romainvo) [inline deepinv/utils/patch_extractor.py]: I would be a little more specific here. It seems that in practice, when `pad_if_need=False`, extracted patches still have the requested `patch_size` but the method crops `(img_size - patch_size) % stride` number of pixels from the right and the bottom of the original image, so that the produced patches cover exactly the cropped area.
- thus reconstruction of the original image is not possible when `pad_if_needed=False` and img_size is not compatible with stride and patch_size.

File: deepinv/models/utils.py [hunk 4]
```diff
@@ -68,56 +69,18 @@ def test_pad(model, L, modulo=16):
 
 
 def patchify(
```
Comment (romainvo) [inline deepinv/models/utils.py]: Should we deprecate the alias ? I seems confusing to have two different functions which do exactly the same thing.

Also the same method is also defined in deepinv.utils, I guess it makes more sense to have it there in the global utils than here

File: deepinv/utils/patch_extractor.py [hunk 5]
```diff
@@ -56,3 +74,100 @@ def patch_extractor(
...
+    Input shape: torch.Size([1, 3, 256, 256]), patchified shape: torch.Size([1, 3, 7, 7, 64, 64])
+    >>> list_patch = [patches[0,:, i, j, ...] for i in range(patches.shape[2]) for j in range(patches.shape[3])]
+    >>> dinv.utils.plot([x, make_grid(list_patch, nrow=patches.shape[2])], titles=["Original", "Overlapping patches"])  # doctest: +SKIP
+
+    .. plot::
+
+        import deepinv as dinv
+        from torchvision.utils import make_grid
+        x = dinv.utils.load_example('butterfly.png')
+        patches = dinv.utils.patchify(x, patch_size=64, stride=32)
```
Comment (romainvo) [inline deepinv/utils/patch_extractor.py]: Use `image_to_patches` instead of patchify

File: deepinv/utils/patch_extractor.py [hunk 6]
```diff
@@ -56,3 +74,100 @@ def patch_extractor(
...
+    :return: Patches of shape ``(B, C, n_rows, n_cols, patch_h, patch_w)``.
+
+    |sep|
+
+    :Examples:
+
+    >>> import deepinv as dinv
+    >>> from torchvision.utils import make_grid
+    >>> x = dinv.utils.load_example('butterfly.png')
+    >>> patches = dinv.utils.patchify(x, patch_size=64, stride=32)
```
Comment (romainvo) [inline deepinv/utils/patch_extractor.py]: Use image_to_patches instead of patchify

File: deepinv/utils/patch_extractor.py [hunk 7]
```diff
@@ -56,3 +74,100 @@ def patch_extractor(
...
+    _, stride_2d = _resolve_tiling_params(patch_size=patches.shape[-2:], stride=stride)
+    return _patches_to_image_impl(
+        patches=patches,
+        stride=stride_2d,
+        img_size=img_size,
+        reduce_overlap=reduce_overlap,
+    )
+
+
+def patchify(
```
Comment (romainvo) [inline deepinv/utils/patch_extractor.py:177]: Same comment that in models/utils.py, should we deprecate the alias ? 

Also we should remove the `patchify` in method/utils.py and only keep this one.

File: deepinv/models/noise_level_estimation.py [hunk 8]
```diff
@@ -1,7 +1,8 @@
+from __future__ import annotations
 import torch
 import torch.nn as nn
 
-from deepinv.models.utils import patchify
+from deepinv.utils.patch_extractor import image_to_patches
```
Comment (romainvo) [inline deepinv/models/noise_level_estimation.py:5]: More of a curiosity: what's the best practice when importing modules internally, should we do `from deepinv.utils import image_to_patches` ? would be less sensitive to future refactoring of the code I guess

File: deepinv/optim/prior.py [hunk 9]
```diff
@@ -528,22 +529,8 @@ def __init__(
...
-            x = torch.cat(
-                (
-                    torch.flip(x[:, :, :, -self.patch_size : -1], (3,)),
-                    x,
-                    torch.flip(x[:, :, :, 1 : self.patch_size], (3,)),
-                ),
-                3,
-            )
+            pad = self.patch_size - 1
+            x = F.pad(x, (pad, pad, pad, pad), mode="reflect")
```
Comment (romainvo) [inline deepinv/optim/prior.py]: We could make the padding mode an option to the user for more flexibility, e.g. as an instance attribute ?

File: deepinv/utils/__init__.py [hunk 10]
```diff
@@ -42,7 +42,12 @@
     dirac_comb_like,
 )
 from .phantoms import RandomPhantomDataset, SheppLoganDataset
-from .patch_extractor import patch_extractor
+from .patch_extractor import (
+    patch_extractor,
+    image_to_patches,
+    patches_to_image,
+    patchify,
```
Comment (romainvo) [inline deepinv/utils/__init__.py:50]: same comment than above about having two methods for the same thing (patchify and image_to_patches)

File: deepinv/utils/patch_extractor.py [hunk 11]
```diff
@@ -56,3 +74,100 @@ def patch_extractor(
     patches = patches.reshape(B, n_patches, C, patch_size, patch_size)
 
     return patches, linear_inds
+
+
+def image_to_patches(
```
Comment (romainvo) [inline deepinv/utils/patch_extractor.py:78]: Why not exposing the `extra_pad` feature of _image_to_patches_impl here like you did in the Mixin class ?

File: deepinv/utils/mixins.py [hunk 12]
```diff
@@ -378,29 +384,29 @@ def image_to_patches(
...
-        patches = image.unfold(2, patch_size[0] + pad[2] + pad[3], stride[0]).unfold(
-            3, patch_size[1] + pad[0] + pad[1], stride[1]
+        patch_size = _add_tuple(self.patch_size, (pad[2] + pad[3], pad[0] + pad[1]))
+        return _image_to_patches_impl(
+            image=image,
+            patch_size=patch_size,
+            stride=self.stride,
+            pad_if_needed=self.pad_if_needed,
+            extra_pad=pad,
         )
```
Comment (romainvo) [inline deepinv/utils/mixins.py]: Should we use the functional call to deepinv.utils.image_to_patches instead, and have the `patch_size = _add_tuple(self.patch_size, (pad[2] + pad[3], pad[0] + pad[1]))` logic only written once in the code ? 
- requires exposing the `extra_pad` feature  in `deepinv.utils.image_to_patches` (with the `pad` kwarg like you do here)

File: deepinv/utils/mixins.py [hunk 13]
```diff
@@ -416,56 +422,16 @@ def patches_to_image(
 
         :param torch.Tensor patches: Patches tensor of shape `(B, C, n_rows, n_cols, patch_h, patch_w)`.
         :param img_size: Target output size (height, width). If provided, output is cropped to this size from the top-left corner.
-        :param reduce_overlap: How to handle overlapping regions. Options are `"sum"` or `"mean"`.
+        :param reduce_overlap: How to handle overlapping regions. Options are `"sum"` or `"mean"`. If `"sum"`, overlapping regions are summed (default). If `"mean"`, overlapping regions are averaged, which can give an identical reconstruction to the original image.
```
Comment (romainvo) [inline deepinv/utils/mixins.py:405]: same comment, why do `"sum"` by default and not `"mean"`

**Review by romainvo — CHANGES_REQUESTED** [review]
Thanks @mh-nguyen712 for the refactoring :)!  Mainly comments about docs and where to put code/logic to avoid duplicates.

----

Some suggested change that I could not add in the review:
In the `TiledMixin2d` init, we should use `_resolve_tiling_params` to avoid reimplementing the same logic twice

File: deepinv/utils/patch_extractor.py (hunk 1 above)
Comment (romainvo) [reply deepinv/utils/patch_extractor.py]: Just saw that you use summation in TiledSpaceVaryingBlur by default, sorry, I was in inference-by-patch mode, feel free to ignore

File: deepinv/utils/patch_extractor.py (hunk 3 above)
Comment (romainvo) [reply deepinv/utils/patch_extractor.py]: In this snippet for example, patches have shape `torch.Size([1, 3, 5, 5, 64, 64])`

```python
import deepinv as dinv
from torchvision.utils import make_grid

x = dinv.utils.load_example('butterfly.png', img_size=(217, 217))

patches = dinv.utils.image_to_patches(x, patch_size=64, stride=32, pad_if_needed=False)
print(patches.shape)

img_from_patches = dinv.utils.patches_to_image(patches, stride=32, img_size=x.shape[-2:], reduce_overlap="mean")
list_patch = [patches[0, :, i, j, ...] for i in range(patches.shape[2]) for j in range(patches.shape[3])]
dinv.utils.plot([x, make_grid(list_patch, nrow=patches.shape[2]), img_from_patches], titles=["Original", "Overlapping patches", "Reconstructed"])
```

Capture d’écran 2026-04-20 201043

File: deepinv/models/noise_level_estimation.py (hunk 8 above)
Comment (mh-nguyen712) [reply deepinv/models/noise_level_estimation.py:5]: from my experience, this is to avoid potential circular import since a full path import will impact less function/module

File: deepinv/models/utils.py (hunk 4 above)
Comment (mh-nguyen712) [reply deepinv/models/utils.py]: Yes I agree about putting it in deepinv.utils. I just kept it for backward-compatibility, but can remove or move to deepinv.utils
I think having alias is not problematic, as some people would prefer the name `patchify` :D 
I move this to deepinv.utils?

File: deepinv/optim/prior.py (hunk 9 above)
Comment (mh-nguyen712) [reply deepinv/optim/prior.py]: Done!

File: deepinv/models/utils.py (hunk 4 above)
Comment (mh-nguyen712) [reply deepinv/models/utils.py]: Removed it!

File: deepinv/utils/patch_extractor.py (hunk 7 above)
Comment (mh-nguyen712) [reply deepinv/utils/patch_extractor.py:177]: I think we can keep the alias, some people would prefer this naming :D

File: deepinv/utils/patch_extractor.py (hunk 1 above)
Comment (mh-nguyen712) [reply deepinv/utils/patch_extractor.py]: Thanks good idea, set default to `mean`

File: deepinv/utils/patch_extractor.py (hunk 3 above)
Comment (mh-nguyen712) [reply deepinv/utils/patch_extractor.py]: Thanks for the remark!

Comment (mh-nguyen712) [reply deepinv/utils/patch_extractor.py]: I clarified a bit more the docs

File: deepinv/utils/mixins.py (hunk 12 above)
Comment (mh-nguyen712) [reply deepinv/utils/mixins.py]: Done

File: deepinv/utils/patch_extractor.py (hunk 11 above)
Comment (mh-nguyen712) [reply deepinv/utils/patch_extractor.py:78]: Done

File: deepinv/utils/patch_extractor.py [hunk 14]
```diff
@@ -56,3 +73,116 @@ def patch_extractor(
...
+    r"""Split a batch of images into overlapping 2D patches.
+
+    The behavior mirrors :class:`deepinv.utils.TiledMixin2d` while exposing a functional API.
+
+    :param torch.Tensor image: Input image tensor of shape ``(B, C, H, W)``.
+    :param int | tuple[int, int] patch_size: Patch size ``(patch_h, patch_w)``.
+        If an ``int`` is provided, the same value is used for both dimensions.
+    :param int | tuple[int, int] stride: Stride between adjacent patches as
+        ``(stride_h, stride_w)``. If ``None``, defaults to half the patch size.
+    :param bool pad_if_needed: If ``True``, adds extra right/bottom padding so that the patches cover the entire image. Default is ``True``.
```
Comment (romainvo) [inline deepinv/utils/patch_extractor.py:94]: Add description for  `pad` kwarg

File: docs/source/changelog.rst [hunk 15]
```diff
@@ -21,7 +21,7 @@ Changed
 - Make available every mask used at evaluation for splitting models :class:`deepinv.loss.SplittingLoss.SplittingModel` when ``eval_n_samples > 1`` (:gh:`881` by `Jérémy Scanvic`_)
 - Deprecate `Loss.name` in favor of the class name as done in the trainer (:gh:`881` by `Jérémy Scanvic`_)
 - Update references for single pixel demo (:gh:`1151` by `Laura C. Diaz-Delgado`_)
-- Add changelog section to the contributing guidelines (:gh:`1153` by `Thibaut Modrzyk`_)
+- Add changelog section to the contributing guidelines (:gh:`1153` by `Thibaut Modrzyk`_)- Unify patching / tiling and unpatching / un-tiling logic in physics and utils, with support for padding and non-overlapping patches. Add :func:`deepinv.utils.image_to_patches` and :func:`deepinv.utils.patches_to_image` utility functions, and refactor physics to use them instead of `unfold` (:gh:`1104` by `Minh Hai Nguyen`_)
```
Comment (romainvo) [inline docs/source/changelog.rst]:
```suggestion
- Add changelog section to the contributing guidelines (:gh:`1153` by `Thibaut Modrzyk`_)
- Unify patching / tiling and unpatching / un-tiling logic in physics and utils, with support for padding and non-overlapping patches. Add :func:`deepinv.utils.image_to_patches` and :func:`deepinv.utils.patches_to_image` utility functions, and refactor physics to use them instead of `unfold` (:gh:`1104` by `Minh Hai Nguyen`_)
```

File: deepinv/physics/blur.py [hunk 16]
```diff
@@ -1070,7 +1070,11 @@ def A_adjoint(
         y = self._pad(y, pad)
 
         # Extract patches
+        print(
+            f"DEBUG -- A_adjoint: y shape after padding: {y.shape}, expected pad: {pad}"
+        )
         patches = self.image_to_patches(y)
+        print(f"DEBUG -- A_adjoint: patches shape after extraction: {patches.shape}")
```
Comment (romainvo) [inline deepinv/physics/blur.py]: Clean debug prints

```suggestion
        patches = self.image_to_patches(y)
```

**Review by romainvo — COMMENTED** [review]
Thanks for the changes! Last few comments on small doc and remaining debug prints.

A last point about refactoring: I think we should factorize the logic in `TiledSpaceVaryingBlur.num_filters` and `_compute_needed_pad`, as they essentially do the same thing. You could make `_compute_needed_pad` return `n_h` and `n_w` for example and afterwards use `_compute_needed_pad` in the `num_filters` helper

File: docs/source/changelog.rst (hunk 15 above)
Comment (mh-nguyen712) [reply docs/source/changelog.rst]: Thanks!

File: deepinv/utils/patch_extractor.py (hunk 14 above)
Comment (mh-nguyen712) [reply deepinv/utils/patch_extractor.py:94]: done, thanks for the remark!

**Comment by mh-nguyen712** [comment]
Thanks again for the careful review @romainvo! I made the refactoring and removed debugging code

**Review by romainvo — APPROVED** [review]
Thanks for the final changes @mh-nguyen712 ! Looks good to me :)
